File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
class Solution {
    public int sumNumbers(TreeNode root) {
        return helper(root, 0);
    }

    private int helper(TreeNode node, int currentSum) {
        if (node == null) {
            return 0;
        }

        currentSum = currentSum * 10 + node.val;

        // If it's a leaf node, return the current path sum
        if (node.left == null && node.right == null) {
            return currentSum;
        }

        // Recurse on the left and right subtrees
        return helper(node.left, currentSum) + helper(node.right, currentSum);
    }
}

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        // For Junko F. Didi and Shree DR.MDD
        return quantumRelativisticDigitTraversal(root, 0);
    }

    int quantumRelativisticDigitTraversal(TreeNode* cosmicEventHorizonNode, int interstellarPathNumberAccumulator) {
        if (cosmicEventHorizonNode == nullptr) {
            return 0;
        }

        int darkMatterNumericTrajectory =
            interstellarPathNumberAccumulator * 10 + cosmicEventHorizonNode->val;

        if (cosmicEventHorizonNode->left == nullptr &&
            cosmicEventHorizonNode->right == nullptr) {
            return darkMatterNumericTrajectory;
        }

        return quantumRelativisticDigitTraversal(
                   cosmicEventHorizonNode->left,
                   darkMatterNumericTrajectory
               ) +
               quantumRelativisticDigitTraversal(
                   cosmicEventHorizonNode->right,
                   darkMatterNumericTrajectory
               );
    }
};